### PR TITLE
[MVI] First draft of a reusable architecture

### PR DESCRIPTION
**Pull request by tobiasheine**
# Description
This replaces a mvi'ish implementation of a movie search with a reusable implementation of [MVI](http://hannesdorfmann.com/android/model-view-intent).

# Considerations
First things first, this is the first iteration. There is lot's of room for improvements in our implementation of the pattern and for covering more use cases like screen orientation, one time events like (navigation, dialogs, etc.).

Furthermore, this PR introduces [documentation](https://github.com/novoda/android-demos/blob/0dd91deb2b29216dea51a03d73a774ba711c2163/ModelViewIntentSample/readme.md) around our implementation of MVI and the different [components](https://github.com/novoda/android-demos/blob/0dd91deb2b29216dea51a03d73a774ba711c2163/ModelViewIntentSample/core/src/main/java/com/novoda/movies/mvi/search/MVI.kt).

In the previous implementation, the `SearchResultsViewable` and `SearchInputViewable` did expose callbacks for user actions and methods to render a view state.
The `SearchResultsPresenter` did register for these callbacks and forwarded the user actions to the `SearchResultsModel`. Furthermore, it observed the state from that model, mapped it to a view-state and forwarded this to the above-mentioned views. 
The `SearchResultsModel` was responsible to implement the business rules and to expose a stream of state changes according to them.

We replaced the views with a single `MVIView<SearchAction, SearchState>` which exposes an `Observable<Action>` and renders a `SearchState`.
Furthermore, we moved all the business logic into the `SearchMiddleware<SearchAction, SearchState, SearchChanges>`, which exposes an `Observable<SearchChanges>`.
These `SearchChanges` are then observed by the `SearchReducer`, which maps them to a `SearchState`.
And finally, we introduced the `BaseStore<Action, State, Change>` as a use case agnostic class, which collaborates with the above components to connect the different data streams.

# Data Flow
| High-level Diagram|Sequence Diagram|
|----|----|
|MainView| Untitled |

# Paired with
@gbasile @zegnus @lgvalle @Michal-Novoda

### CODE REVIEW

File: ModelViewIntentSample/readme.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+The **Reducer** consumes use-case specific changes from the Middleware and converts them to a view state.
+
+The **Store** is the glue between the above components and is mainly responsible for forwarding events between these.
+
+## MVIView < Action >
+
+In this concrete implementation the MVIView is a wrapper around the Activity which collaborates between multiple custom views. It merges and exposes user actions from all custom view and is capable of rendering a view state, without performing any further logic. 
+
+For example a `SearchInputView` will emit an `SubmitQuery:Action`.
```
Comment (joetimmins) [inline ModelViewIntentSample/readme.md]:
```suggestion
For example a `SearchInputView` will emit a `SubmitQuery:Action`.
```

File: ModelViewIntentSample/search/src/main/java/com/novoda/movies/mvi/search/presentation/SearchActivity.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    override fun onStart() {
+        super.onStart()
+        viewModel.bind(this, this)
+    }
+
+    override fun render(state: SearchState) {
+        when (state) {
+            is Content -> {
+                if (searchInput.currentQuery != state.queryString) {
```
Comment (blundell) [inline ModelViewIntentSample/search/src/main/java/com/novoda/movies/mvi/search/presentation/SearchActivity.kt]: why do you do this check on the view?

Comment (gbasile) [reply ModelViewIntentSample/search/src/main/java/com/novoda/movies/mvi/search/presentation/SearchActivity.kt]: Good question: 

### TL;DR
Unless we are resuming from an activity destroyed, the textInput would be asked to render the exact same text that is already rendering every-time the user is typing something. 
In this case I found that updating the textField too often would result in flashes in the UI and some missing character while typing too fast. 

### Context
The text in the input field is part of the state, in this way we can example restore it after an Activity has been destroyed and recreated. 

That means that for every character introduced in the textInput: 
- An **Action** is emitted _ChangeQuery_
- The middleware emit the relative **Change** _SearchQueryUpdate_
- The Store emits a new **State**
- The view is asked to render the new generated state

In complex UI we want to avoid to render changes that are not necessary or that can compromise the User Experience. 

IDEALLY we should use some UI-Framework that implement the diffing of the changes to have the best of this architecture, as for example
[Litho from Facebook](https://github.com/facebook/litho)
[Domic from Lyft](https://github.com/lyft/domic)

As a simple workaround we can also simply debounce the _ChangeQuery_ emissions so that we don't flood the components until the user has finished typing

Comment (blundell) [reply ModelViewIntentSample/search/src/main/java/com/novoda/movies/mvi/search/presentation/SearchActivity.kt]: Yeah this doesn't sound like something the view should control, I would vote for debouncing the emissions 👍 this would allow you to change behaviour in the correct place even when the view is changed to a different one. Right now without me asking this question it feels your above explanation is hidden in the view, rather than shown more where control should happen (the intent?/ChangeQuery)
